pkg/internal/docker/start_jenkins.go

```diff
@@ -0,0 +1,159 @@
+				{
+					HostIP:   "0.0.0.0",
+					HostPort: strconv.Itoa(o.JenkinsPort),
+				},
+			},
+		},
+	}
+	exposedPorts := map[nat.Port]struct{}{
+		jenkinsPort: struct{}{},
+	}
+	config := &container.Config{
+		Image:        imageName,
+		ExposedPorts: exposedPorts,
+	}
+
+	resp, err := cli.ContainerCreate(ctx, config, hostConfig, nil, nil, "jenkinstest2")
+	if err != nil {
+		fmt.Println(err)
+	}
+	cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
+	fmt.Println(resp.ID)
+	return nil
+}
+
+//BuildImage is used to build an image
+func (o *RunOptions) BuildImage(cmd *cobra.Command) error {
+	ctx := context.Background()
+	cli, _ := o.ConnectToDocker()
+	dockerFileTarReader, _ := o.GetTarReader(cmd)
+	buildOptions := types.ImageBuildOptions{
+		Context:    dockerFileTarReader,
+		Dockerfile: o.DockerfilePath,
+		Remove:     true,
+		Tags:       []string{o.ImageName},
+	}
+	imageBuildResponse, err := cli.ImageBuild(
+		ctx,
+		dockerFileTarReader,
+		buildOptions,
+	)
+
+	if err != nil {
+		return err
+	}
+	defer imageBuildResponse.Body.Close()
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(imageBuildResponse.Body)
+	cmd.Println(buf.String())
+	return nil
+}
+
+//GetTarReader creates tarReader for args in function BuildImage
+func (o *RunOptions) GetTarReader(cmd *cobra.Command) (*bytes.Reader, error) {
+	src := []string{o.DockerfilePath, o.WarPath, "/tmp/jenkins-cli-docker.sh"}
+	buf := new(bytes.Buffer)
+	tw := tar.NewWriter(buf)
+	defer tw.Close()
+
+	for _, fileName := range src {
+		dockerFileReader, err := os.Open(fileName)
+		if err != nil {
+			return nil, err
+		}
+		readDockerFile, err := ioutil.ReadAll(dockerFileReader)
+		if err != nil {
+			return nil, err
+		}
+		tarHeader := &tar.Header{
+			Name: fileName,
+			Size: int64(len(readDockerFile)),
+		}
+		err = tw.WriteHeader(tarHeader)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tw.Write(readDockerFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	dockerFileTarReader := bytes.NewReader(buf.Bytes())
+	return dockerFileTarReader, nil
+}
+
+//CreateDockerfile will create a docker file for running a jenkins war contains plugins hpi
+func (o *RunOptions) CreateDockerfile(cmd *cobra.Command, args []string) (err error) {
+	sh := "java -Djenkins.install.runSetupWizard=false -jar /usr/share/jenkins/jenkins.war\n" +
+		"tail -f /dev/null"
+	err = ioutil.WriteFile("/tmp/jenkins-cli-docker.sh", []byte(sh), 0)
+	if err != nil {
+		panic(err)
+	}
+	dockerfileString := fmt.Sprintf("FROM adoptopenjdk/openjdk11\n"+
+		"LABEL Version=\"1.0-SNAPSHOT\"Description=\"Jenkins formula generated by jcli\"Vendor=\"Chinese Jenkins Community\"\n"+
+		"ADD %s /usr/share/jenkins/jenkins.war\n"+
+		"ADD /tmp/jenkins-cli-docker.sh /usr/local/bin/jenkins-cli-docker.sh\n"+
+		"ENTRYPOINT [\"sh\", \"/usr/local/bin/jenkins-cli-docker.sh\"]", o.WarPath)
+	err = ioutil.WriteFile("/tmp/Dockerfile", []byte(dockerfileString), 0)
+	if err != nil {
+		panic(err)
+	}
+	return nil
+}
```

Comment:
This function looks not used.

Reply:
I plan to use this function separately and this function needs to run before 'CreateImageAndRunContainer'.